app/routes/api+/custom-fields.bulk-actions.ts

```diff
@@ -0,0 +1,158 @@
+import { json, type ActionFunctionArgs } from "@remix-run/node";
+import { z } from "zod";
+import { BulkActivateCustomFieldSchema } from "~/components/custom-fields/bulk-activate-dialog";
+import { BulkDeactivateCustomFieldSchema } from "~/components/custom-fields/bulk-deactivate-dialog";
+import { db } from "~/database/db.server";
+import {
+  bulkActivateOrDeactivateCustomFields,
+  countActiveCustomFields,
+} from "~/modules/custom-field/service.server";
+import { getOrganizationTierLimit } from "~/modules/tier/service.server";
+import { checkExhaustiveSwitch } from "~/utils/check-exhaustive-switch";
+import { sendNotification } from "~/utils/emitter/send-notification.server";
+import { makeShelfError, ShelfError } from "~/utils/error";
+import { assertIsPost, data, error, parseData } from "~/utils/http.server";
+import { ALL_SELECTED_KEY } from "~/utils/list";
+import {
+  PermissionAction,
+  PermissionEntity,
+} from "~/utils/permissions/permission.validator.server";
+import { requirePermission } from "~/utils/roles.server";
+import {
+  canCreateMoreCustomFields,
+  willExceedCustomFieldLimit,
+} from "~/utils/subscription.server";
+
+export async function action({ request, context }: ActionFunctionArgs) {
+  const authSession = context.getSession();
+  const { userId } = authSession;
+
+  try {
+    assertIsPost(request);
+
+    const formData = await request.formData();
+
+    const { intent } = parseData(
+      formData,
+      z.object({ intent: z.enum(["bulk-activate", "bulk-deactivate"]) })
+    );
+
+    const intentToActionMap: Record<typeof intent, PermissionAction> = {
+      "bulk-activate": PermissionAction.update,
+      "bulk-deactivate": PermissionAction.update,
+    };
+
+    const { organizationId, organizations } = await requirePermission({
+      userId,
+      request,
+      entity: PermissionEntity.customField,
+      action: intentToActionMap[intent],
+    });
+
+    switch (intent) {
+      case "bulk-activate": {
+        const { customFieldIds } = parseData(
+          formData,
+          BulkActivateCustomFieldSchema
+        );
+
+        const newActivatingFields = await db.customField.count({
+          where: customFieldIds.includes(ALL_SELECTED_KEY)
+            ? { organizationId }
+            : { id: { in: customFieldIds } },
+        });
+
+        /** We have to make sure that user is not already at the limit */
+        const tierLimit = await getOrganizationTierLimit({
+          organizationId,
+          organizations,
+        });
+
+        const totalActiveCustomFields = await countActiveCustomFields({
+          organizationId,
+        });
+
+        const canCreateMore = canCreateMoreCustomFields({
```

Comment:
Inside the actions/loaders you should use `assertUserCanCreateMoreCustomFields`. That already throws an error. We use `canCreateMoreCustomFields` only when we need a boolean value to pass to the view.